### ISSUE TITLE
Some tests for no flux boundary conditions, diffusion, and Boussinesq.

### ISSUE THREAD

**Issue by ali-ramadhan** [climate-machine/Oceananigans.jl#126]
@glwagner and @SandreOuza suggested the following tests. We should implement these using the new boundary conditions API #118.

No flux tests:
1. Set _T(t=0) = 1_ and make sure it stays the same.
2. Set _T(t=0) = cos(z) + 1/2_ and _∂T/∂z =_ and make sure you get _T(t) -> 1/2_.

Diffusion test:
* Set _T(t=0)_ to be a linear temperature profile with slope _c_ and impose that _∂T/∂z = c_ at the top and bottom. Make sure this stably stratified temperature profile is maintained.

Boussinesq test:
* Free convection test: set a cooling flux _Q_ at the surface and a stratification of _N²_. Make sure the mixed layer deepens at the correct rate.

Related to #81.

**Comment by ali-ramadhan** [comment]
Photo with notes on parameters used for free convection test. Write up by @SandreOuza, photo by @christophernhill. @jm-c is running this test using the MITgcm. GitHub changed the photo orientation :(

![IMG-4239](https://user-images.githubusercontent.com/20099589/54446927-7cce5a00-471e-11e9-8633-8ddbfc155aff.JPG)

**Comment by ali-ramadhan** [comment]
Notes on the diffusion test:
```julia
Nx, Ny, Nz = 16, 16, 16
Lx, Ly, Lz = 100, 100, 100
Δt = 1
ν, κ = 1e-2, 1e-2

model = Model(N=(Nx, Ny, Nz), L=(Lx, Ly, Lz), ν=ν, κ=κ,  arch=:CPU,  float_type=Float64)
model.boundary_conditions = BoundaryConditions(:periodic, :periodic, :rigid_lid, :free_slip)

T_prof = model.eos.T₀ .+ 0.07 .* model.grid.zC
T_3d   = repeat(reshape(T_prof, 1, 1, Nz), Nx, Ny, 1)

model.tracers.T.data .= T_3d

top_flux(args...) = ifelse(k == 1,  c*κv / Δz, 0)
bottom_flux(args...) = ifelse(k == Nz,  -c*κv / Δz, 0)

model.forcings = (nothing, nothing, nothing, top_flux + bottom_flux, nothing)

for i in 1:10
    time_step!(model, 1, Δt)
    @test all(model.G.GT.data[1, 1, 1] .≈ 0)
    @test all(model.G.GT.data[1, 1, end] .≈ 0)
end
```

**Comment by ali-ramadhan** [comment]
Fixed in #118.
